CheckNotificationThread

BekbolotGM opened the pull request:
Функция CheckNotificationThread() уведомляет пользователся не написавшего стендап 3 раза каждые 30 мин.

Commit:
Implement function ExecuteThread()

Commit:
Implement aditional db methods and write tests

Commit:
Make changes in tests

Commit:
Change function EditDeadline()

Commit:
Delete extra lines

Commit:
Implement function CheckNotificationThread()

BekbolotGM committed:
Merge branch 'master' into execThread

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -14,7 +15,11 @@ import (
-const warnPeriod = 10 // 10 minutes before the deadline
+const warnPeriod = 10       // 10 minutes before the deadline
+const notificationTime = 30 // 30 minutes after deadline
+const maxReminder = 3
```
Перемести константы в конфиг хотябы. в идеале это вообще должны быть настройки группы. так как каждый сам волен указывать сколько раз напоминать, через сколько минут и т.д.

anatoliyfedorenko committed:
NEED FIXES!

Commit:
Fix db methods and write tests

Commit:
Fix code
Change model NotificationThread, write tests
Change migrations file 00012_notifications_thread
Change function CheckNotificationThread()

Commit:
Delete global variable

Commit:
Change value of notification time

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,81 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+			err = b.db.DeleteNotificationThread(thread.ID)
+			if err != nil {
+				log.Error("Error on executing DeleteNotificationsThread! ", err)
+				return
```
должно быть continue

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,81 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+	if strings.TrimSpace(group.StandupDeadline) == "" {
+		threads, err := b.db.ListNotificationsThread(group.ChatID)
+		if err != nil {
+			log.Error("Error on executing CheckNotificationThread! ", err)
```
лог потом не найдём. поправить

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,81 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+		for _, thread := range threads {
+			err = b.db.DeleteNotificationThread(thread.ID)
+			if err != nil {
+				log.Error("Error on executing DeleteNotificationsThread! ", err)
```
укажи хоть какой это трэд, чтобы понять ,что пошло не так.

BekbolotGM replied:
То есть ID треда?

Commit:
Fix code in the file notifications.go

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -196,8 +199,17 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+			ReminderCounter:  0,
+		})
+		if err != nil {
+			log.Error("Error on executing CreateNotificationThread ", err)
```
тоже надо вставить что-нибдуь чтобы понять, с какими именно параметрами функция упала.

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,79 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+	if strings.TrimSpace(group.StandupDeadline) == "" {
+		threads, err := b.db.ListNotificationsThread(group.ChatID)
+		if err != nil {
+			log.Error("Error on executing ListNotificationsThread! ", err)
```
тоже надо вставить что-нибудь чтобы понять, с какими именно параметрами функция упала.

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,79 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+
+	threads, err := b.db.ListNotificationsThread(group.ChatID)
+	if err != nil {
+		log.Error("Error on executing ListNotificationThread! ", err)
```
вставить group.ChatID в логгер. Везде оставляй конкретные данные, чтобы понять больше. когда будешь дебажить, поблагодаришь

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,79 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+
+	for _, thread := range threads {
+		loc, err := time.LoadLocation(group.TZ)
+		if t.Hour() == thread.NotificationTime.In(loc).Hour() || t.Minute() == thread.NotificationTime.In(loc).Minute() {
```
у тебя очень большая матрёшка из if условий. Можно было сделать так, что если время не равно времени треда, то просто пропускать, а в других случаях делать код, который ниже. Чем больше у тебя вложенных условий, тем тяжелее код читать.

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,79 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+			notify, err := localizer.Localize(&i18n.LocalizeConfig{
+				DefaultMessage: &i18n.Message{
+					ID:    "remindNonReporter",
+					Other: "Attention, @{{.Intern}}! you still haven't written a standup! Write a standup!",
```
Замени здесь тоже слово intern. по всему коду нужно от этого избавиться. можно просто standuper , так как бот больше не для стажёров

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -230,3 +242,79 @@ func (b *Bot) NotifyGroup(group *model.Group, t time.Time) {
+			msg.ParseMode = "Markdown"
+			_, err = b.tgAPI.Send(msg)
+			if err != nil {
+				log.Error("Error sending message, CheckNotificationThread! ", err)
```
Тоже нужны доп параметры! Не поймёшь же в какую группу отправлял, запутаешься в логах.

anatoliyfedorenko commented on config/config.go:
```diff
@@ -1,16 +1,23 @@
+	TelegramToken    string `envconfig:"TELEGRAM_TOKEN" required:"false"`
+	DatabaseURL      string `envconfig:"DATABASE_URL" required:"false" default:"telegram:telegram@tcp(localhost:3306)/telegram?parseTime=true"`
+	MaxReminder      int
+	NotificationTime int
```
А тэги для MaxReminders и NotificationTime ты решил не прописывать? пример то, что выше, тебя вообще никак не насторожил да? Там ведь и default есть (значение по умолчанию) и envconfig ( ты ведь вообще не разобрался что это и у меня тоже не уточнил, нужно ли что-либо с этим делать. не почитал доку.

anatoliyfedorenko commented on config/config.go:
```diff
@@ -1,16 +1,23 @@
 // Get config data from environment
 func Get() (*BotConfig, error) {
-	var c BotConfig
+	var c = BotConfig{
```
Убери эту порнографию и лучше в тэги вставь default значения...

Commit:
Fix code
Add default value for MaxReminder and NotificationTime in config.go
Change some messages in active.en.toml and active.tu.toml file
Fix code in notifications.go

Commit:
Delete extra line

anatoliyfedorenko commented on bot/notifications.go:
```diff
@@ -263,58 +263,60 @@ func (b *Bot) CheckNotificationThread(group *model.Group, t time.Time) {
-					continue
-				}
-			}
+		if t.Hour() != thread.NotificationTime.In(loc).Hour() || t.Minute() != thread.NotificationTime.In(loc).Minute() {
```
исправить. не надо так делать ))

Commit:
Fix bug

anatoliyfedorenko commented on config/config.go:
```diff
@@ -1,11 +1,15 @@
-	DatabaseURL   string `envconfig:"DATABASE_URL" required:"false" default:"telegram:telegram@tcp(localhost:3306)/telegram?parseTime=true"`
+	TelegramToken    string `envconfig:"TELEGRAM_TOKEN" required:"false"`
+	DatabaseURL      string `envconfig:"DATABASE_URL" required:"false" default:"telegram:telegram@tcp(localhost:3306)/telegram?parseTime=true"`
+	MaxReminder      int    `default:"3"`
```
добавь envconfig и как должно называться в .env файле этот параметр.

Commit:
Add tag envconfig in config.go
